{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyNqcjkC0xIxTaxARfZIAER/",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/katha815/ADME-Classifier/blob/main/Dataset_Expansion5_autoUnlisted21.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "m_9HrF20mF_F",
        "outputId": "771f4b94-a1ae-47a5-d387-decaa004c480"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/gdrive\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/gdrive')\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt"
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Data preparation (Trial)\n",
        "\n",
        "# List of TSV file paths\n",
        "file_paths = ['/content/gdrive/My Drive/Colab Notebooks/ADME Project/ABCB1.tsv',\n",
        "         '/content/gdrive/My Drive/Colab Notebooks/ADME Project/ABCG2.tsv',\n",
        "         '/content/gdrive/My Drive/Colab Notebooks/ADME Project/CYP1A1.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/CYP2A6.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/CYP2B6.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/CYP2E1.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/CYP3A5.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/DPYD.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/GSTM1.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/GSTP1.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/GDTT1.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/NAT2.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/SLC15A2.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/SLC22A1.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/SLC22A6.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/SLCO1B3.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/SULT1A1.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/UGT1A1.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/UGT2B15.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/UGT2B17.tsv',\n",
        "        '/content/gdrive/My Drive/Colab Notebooks/ADME Project/UGT2B7.tsv']\n",
        "\n",
        "# Create an empty list to store DataFrames\n",
        "dfs = []\n",
        "\n",
        "# Read each TSV file and append its DataFrame to the list\n",
        "for file_path in file_paths:\n",
        "    try:\n",
        "        df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
        "        dfs.append(df)\n",
        "    except pd.errors.ParserError:\n",
        "        print(f\"Error reading file: {file_path}\")\n",
        "\n",
        "# Concatenate the DataFrames into a single DataFrame\n",
        "comb_df = pd.concat(dfs, ignore_index=True)\n",
        "\n",
        "# Save the combined DataFrame as a CSV file\n",
        "comb_df.to_csv('/content/gdrive/My Drive/Colab Notebooks/ADME Project/unlisted21_auto.csv', index=False)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "_yj9FDyjeGlk",
        "outputId": "4e2095f5-c6df-47fe-8118-d41c56c59954"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-82-b23c12e0001c>:32: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Results are reliable: same data size with manual version!"
      ],
      "metadata": {
        "id": "3LNeQmUmjQtp"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "Core_genes = pd.read_csv(\"/content/gdrive/My Drive/Colab Notebooks/ADME Project/Core ADME Genes.csv\")"
      ],
      "metadata": {
        "id": "a-W2aKLLmeo1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "Core_genes['Gene'][0]\n",
        "gene_name = Core_genes['Gene'][0]\n",
        "gene_name"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 36
        },
        "id": "5ASgFlnOnO0w",
        "outputId": "d52591f0-f407-4d62-f789-99adaa36097e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "'ABCB1'"
            ],
            "application/vnd.google.colaboratory.intrinsic+json": {
              "type": "string"
            }
          },
          "metadata": {},
          "execution_count": 84
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Data preparation (real run)\n",
        "file_paths = []\n",
        "\n",
        "# List of TSV file paths\n",
        "for i in range(len(Core_genes['Gene'])):\n",
        "  if Core_genes['Listed?'][i]==1:\n",
        "    gene_name = Core_genes['Gene'][i]\n",
        "    file_paths.append(f'/content/gdrive/My Drive/Colab Notebooks/ADME Project/{gene_name}.tsv')\n",
        "\n",
        "# Create an empty list to store DataFrames\n",
        "dfs = []\n",
        "\n",
        "# Read each TSV file and append its DataFrame to the list\n",
        "for file_path in file_paths:\n",
        "    try:\n",
        "        df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
        "        dfs.append(df)\n",
        "    except pd.errors.ParserError:\n",
        "        print(f\"Error reading file: {file_path}\")\n",
        "\n",
        "# Concatenate the DataFrames into a single DataFrame\n",
        "raw_df = pd.concat(dfs, ignore_index=True)\n",
        "\n",
        "# Save the combined DataFrame as a CSV file\n",
        "raw_df.to_csv('/content/gdrive/My Drive/Colab Notebooks/ADME Project/listedCore_auto.csv', index=False)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 235
        },
        "id": "q_9EKEWQjgTL",
        "outputId": "c0d1562c-5031-4896-91ef-d5f8fd5793d4"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "error",
          "ename": "NameError",
          "evalue": "ignored",
          "traceback": [
            "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
            "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
            "\u001b[0;32m<ipython-input-3-1b968528f97f>\u001b[0m in \u001b[0;36m<cell line: 5>\u001b[0;34m()\u001b[0m\n\u001b[1;32m      3\u001b[0m \u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      4\u001b[0m \u001b[0;31m# List of TSV file paths\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0;32m----> 5\u001b[0;31m \u001b[0;32mfor\u001b[0m \u001b[0mi\u001b[0m \u001b[0;32min\u001b[0m \u001b[0mrange\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mlen\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0mCore_genes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Gene'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m\u001b[1;32m      6\u001b[0m   \u001b[0;32mif\u001b[0m \u001b[0mCore_genes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Listed?'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m==\u001b[0m\u001b[0;36m1\u001b[0m\u001b[0;34m:\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[1;32m      7\u001b[0m     \u001b[0mgene_name\u001b[0m \u001b[0;34m=\u001b[0m \u001b[0mCore_genes\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m'Gene'\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0mi\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n",
            "\u001b[0;31mNameError\u001b[0m: name 'Core_genes' is not defined"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Ref/Alter rows creation (Trial)\n",
        "df = pd.DataFrame({'Alleles': ['Allele Aasdasd', 'Allele C T', 'allele A G', 'sadasd']})\n",
        "\n",
        "# Define the search patterns and corresponding replacement values\n",
        "patterns = ['Allele A ', 'Allele C ', 'Allele G ', 'Allele T ']\n",
        "patterns2 = ['allele A', 'allele C', 'allele G', 'allele T']\n",
        "replacements = ['A', 'C', 'G', 'T']\n",
        "\n",
        "# Create the two new columns using the search patterns and replacements\n",
        "df['New_Column_1'] = df['Alleles'].apply(lambda x: next((replacements[i] for i, pattern in enumerate(patterns) if pattern in x), np.nan))\n",
        "df['New_Column_2'] = df['Alleles'].apply(lambda x: next((replacements[i] for i, pattern2 in enumerate(patterns2) if pattern2 in x), np.nan))\n",
        "\n",
        "# Display the resulting DataFrame\n",
        "df"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 175
        },
        "id": "S9ZoxmnwrKj9",
        "outputId": "7331d1dc-905a-413d-be99-784bc0fec66d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "          Alleles New_Column_1 New_Column_2\n",
              "0  Allele Aasdasd          NaN          NaN\n",
              "1      Allele C T            C          NaN\n",
              "2      allele A G          NaN            A\n",
              "3          sadasd          NaN          NaN"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-388d3a54-9b9e-48e7-a173-5e2565e27dba\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Alleles</th>\n",
              "      <th>New_Column_1</th>\n",
              "      <th>New_Column_2</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>Allele Aasdasd</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>Allele C T</td>\n",
              "      <td>C</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>allele A G</td>\n",
              "      <td>NaN</td>\n",
              "      <td>A</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>sadasd</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-388d3a54-9b9e-48e7-a173-5e2565e27dba')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-388d3a54-9b9e-48e7-a173-5e2565e27dba button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-388d3a54-9b9e-48e7-a173-5e2565e27dba');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 86
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Ref/Alter rows creation (Real-run)\n",
        "data = comb_df\n",
        "#raw_df\n",
        "\n",
        "# Define the search patterns and corresponding replacement values\n",
        "patterns = ['Allele A ', 'Allele C ', 'Allele G ', 'Allele T ']\n",
        "patterns2 = ['allele A', 'allele C', 'allele G', 'allele T']\n",
        "replacements = ['A', 'C', 'G', 'T']\n",
        "\n",
        "# Create the two new columns using the search patterns and replacements\n",
        "data['Alter'] = data['Association'].apply(lambda x: next((replacements[i] for i, pattern in enumerate(patterns) if pattern in x), np.nan))\n",
        "data['Ref'] = data['Association'].apply(lambda x: next((replacements[i] for i, pattern2 in enumerate(patterns2) if pattern2 in x), np.nan))\n",
        "\n",
        "data"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "80CYdt8-r3OL",
        "outputId": "bb8029e3-427a-4744-a9d7-f4cd95927dcb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "      PharmGKB ID     Variant        Literature                        Drugs  \\\n",
              "0      1183491722   rs1045642     PMID:23900887                   tacrolimus   \n",
              "1      1451335298   rs4728709     PMID:33378980                   olanzapine   \n",
              "2      1183491726   rs1045642     PMID:22725663                    talinolol   \n",
              "3      1451335303   rs7787082     PMID:33378980                   olanzapine   \n",
              "4      1184349841   rs1045642  PMCID:PMC4104276                  clopidogrel   \n",
              "...           ...         ...               ...                          ...   \n",
              "5261   1448098981  rs62298861     PMID:26213157                  lamotrigine   \n",
              "5262    827812768  rs28365062     PMID:21806385                valproic acid   \n",
              "5263   1452121460  rs73823859     PMID:37248824  codeine; morphine; tramadol   \n",
              "5264   1448531887  rs28365063     PMID:28063968                     morphine   \n",
              "5265   1448531882   rs7438135     PMID:28063968                     morphine   \n",
              "\n",
              "                                            Association Significance  \\\n",
              "0     Genotype GG is not associated with risk of tra...           no   \n",
              "1     Allele G is not associated with exposure to ol...           no   \n",
              "2     Genotype AA is associated with increased clear...          yes   \n",
              "3     Allele A is not associated with exposure to ol...           no   \n",
              "4     Allele A is not associated with increased risk...           no   \n",
              "...                                                 ...          ...   \n",
              "5261  Genotype AA is not associated with increased c...           no   \n",
              "5262  Genotype GG is not associated with increased d...           no   \n",
              "5263  Allele A is associated with increased Pain whe...          yes   \n",
              "5264  Allele G is not associated with exposure to mo...           no   \n",
              "5265  Allele G is not associated with exposure to mo...           no   \n",
              "\n",
              "       P-Value  # of Cases  # of Controls  \\\n",
              "0       = 0.70        88.0            NaN   \n",
              "1       > 0.05        80.0            NaN   \n",
              "2      = 0.028        18.0            NaN   \n",
              "3       > 0.05        80.0            NaN   \n",
              "4      = 0.046      1632.0            NaN   \n",
              "...        ...         ...            ...   \n",
              "5261   = 0.278       140.0            NaN   \n",
              "5262       NaN       162.0            NaN   \n",
              "5263  = 0.0454       106.0            NaN   \n",
              "5264    = 0.51        22.0            NaN   \n",
              "5265    = 0.32        22.0            NaN   \n",
              "\n",
              "                                 Biogeographical Groups  Phenotype Categories  \\\n",
              "0                                              European              Efficacy   \n",
              "1     Multiple groups, \"36 Caucasians, 42 Latin-Amer...         Metabolism/PK   \n",
              "2                                            East Asian         Metabolism/PK   \n",
              "3     Multiple groups, \"36 Caucasians, 42 Latin-Amer...         Metabolism/PK   \n",
              "4                                              European              Toxicity   \n",
              "...                                                 ...                   ...   \n",
              "5261                                         East Asian         Metabolism/PK   \n",
              "5262                                         East Asian  Dosage,Metabolism/PK   \n",
              "5263                                Sub-Saharan African              Efficacy   \n",
              "5264                                           European         Metabolism/PK   \n",
              "5265                                           European         Metabolism/PK   \n",
              "\n",
              "      Pediatric                                       More Details   Genes  \\\n",
              "0         False  In those with DONOR LIVERS with these genotype...   ABCB1   \n",
              "1         False  No significant difference in a number of pharm...   ABCB1   \n",
              "2         False  Subjects with the AA genotype had increased or...   ABCB1   \n",
              "3         False  No significant difference in a number of pharm...   ABCB1   \n",
              "4         False                                                NaN   ABCB1   \n",
              "...         ...                                                ...     ...   \n",
              "5261      False  The concentration was measured as lamotrigine ...  UGT2B7   \n",
              "5262      False  Variant described as UGT2B7 735A>G (rsID not r...  UGT2B7   \n",
              "5263       True  \"However, UGT2B7 rs73823859 showed a statistic...  UGT2B7   \n",
              "5264      False  Measures assessed were AUC, transit compartmen...  UGT2B7   \n",
              "5265      False  Measures assessed were AUC, transit compartmen...  UGT2B7   \n",
              "\n",
              "     Alter  Ref  numeric  \n",
              "0      NaN  NaN      0.0  \n",
              "1        G    A      0.0  \n",
              "2      NaN  NaN      1.0  \n",
              "3        A    G      0.0  \n",
              "4        A    G      0.0  \n",
              "...    ...  ...      ...  \n",
              "5261   NaN  NaN      0.0  \n",
              "5262   NaN  NaN      0.0  \n",
              "5263     A    G      1.0  \n",
              "5264     G    A      0.0  \n",
              "5265     G    A      0.0  \n",
              "\n",
              "[5266 rows x 17 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-2a6a9d61-1260-40a6-86a3-6b830165b439\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>PharmGKB ID</th>\n",
              "      <th>Variant</th>\n",
              "      <th>Literature</th>\n",
              "      <th>Drugs</th>\n",
              "      <th>Association</th>\n",
              "      <th>Significance</th>\n",
              "      <th>P-Value</th>\n",
              "      <th># of Cases</th>\n",
              "      <th># of Controls</th>\n",
              "      <th>Biogeographical Groups</th>\n",
              "      <th>Phenotype Categories</th>\n",
              "      <th>Pediatric</th>\n",
              "      <th>More Details</th>\n",
              "      <th>Genes</th>\n",
              "      <th>Alter</th>\n",
              "      <th>Ref</th>\n",
              "      <th>numeric</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>1183491722</td>\n",
              "      <td>rs1045642</td>\n",
              "      <td>PMID:23900887</td>\n",
              "      <td>tacrolimus</td>\n",
              "      <td>Genotype GG is not associated with risk of tra...</td>\n",
              "      <td>no</td>\n",
              "      <td>= 0.70</td>\n",
              "      <td>88.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>European</td>\n",
              "      <td>Efficacy</td>\n",
              "      <td>False</td>\n",
              "      <td>In those with DONOR LIVERS with these genotype...</td>\n",
              "      <td>ABCB1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>1451335298</td>\n",
              "      <td>rs4728709</td>\n",
              "      <td>PMID:33378980</td>\n",
              "      <td>olanzapine</td>\n",
              "      <td>Allele G is not associated with exposure to ol...</td>\n",
              "      <td>no</td>\n",
              "      <td>&gt; 0.05</td>\n",
              "      <td>80.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Multiple groups, \"36 Caucasians, 42 Latin-Amer...</td>\n",
              "      <td>Metabolism/PK</td>\n",
              "      <td>False</td>\n",
              "      <td>No significant difference in a number of pharm...</td>\n",
              "      <td>ABCB1</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1183491726</td>\n",
              "      <td>rs1045642</td>\n",
              "      <td>PMID:22725663</td>\n",
              "      <td>talinolol</td>\n",
              "      <td>Genotype AA is associated with increased clear...</td>\n",
              "      <td>yes</td>\n",
              "      <td>= 0.028</td>\n",
              "      <td>18.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>East Asian</td>\n",
              "      <td>Metabolism/PK</td>\n",
              "      <td>False</td>\n",
              "      <td>Subjects with the AA genotype had increased or...</td>\n",
              "      <td>ABCB1</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>1451335303</td>\n",
              "      <td>rs7787082</td>\n",
              "      <td>PMID:33378980</td>\n",
              "      <td>olanzapine</td>\n",
              "      <td>Allele A is not associated with exposure to ol...</td>\n",
              "      <td>no</td>\n",
              "      <td>&gt; 0.05</td>\n",
              "      <td>80.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Multiple groups, \"36 Caucasians, 42 Latin-Amer...</td>\n",
              "      <td>Metabolism/PK</td>\n",
              "      <td>False</td>\n",
              "      <td>No significant difference in a number of pharm...</td>\n",
              "      <td>ABCB1</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>1184349841</td>\n",
              "      <td>rs1045642</td>\n",
              "      <td>PMCID:PMC4104276</td>\n",
              "      <td>clopidogrel</td>\n",
              "      <td>Allele A is not associated with increased risk...</td>\n",
              "      <td>no</td>\n",
              "      <td>= 0.046</td>\n",
              "      <td>1632.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>European</td>\n",
              "      <td>Toxicity</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>ABCB1</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5261</th>\n",
              "      <td>1448098981</td>\n",
              "      <td>rs62298861</td>\n",
              "      <td>PMID:26213157</td>\n",
              "      <td>lamotrigine</td>\n",
              "      <td>Genotype AA is not associated with increased c...</td>\n",
              "      <td>no</td>\n",
              "      <td>= 0.278</td>\n",
              "      <td>140.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>East Asian</td>\n",
              "      <td>Metabolism/PK</td>\n",
              "      <td>False</td>\n",
              "      <td>The concentration was measured as lamotrigine ...</td>\n",
              "      <td>UGT2B7</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5262</th>\n",
              "      <td>827812768</td>\n",
              "      <td>rs28365062</td>\n",
              "      <td>PMID:21806385</td>\n",
              "      <td>valproic acid</td>\n",
              "      <td>Genotype GG is not associated with increased d...</td>\n",
              "      <td>no</td>\n",
              "      <td>NaN</td>\n",
              "      <td>162.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>East Asian</td>\n",
              "      <td>Dosage,Metabolism/PK</td>\n",
              "      <td>False</td>\n",
              "      <td>Variant described as UGT2B7 735A&gt;G (rsID not r...</td>\n",
              "      <td>UGT2B7</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5263</th>\n",
              "      <td>1452121460</td>\n",
              "      <td>rs73823859</td>\n",
              "      <td>PMID:37248824</td>\n",
              "      <td>codeine; morphine; tramadol</td>\n",
              "      <td>Allele A is associated with increased Pain whe...</td>\n",
              "      <td>yes</td>\n",
              "      <td>= 0.0454</td>\n",
              "      <td>106.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Sub-Saharan African</td>\n",
              "      <td>Efficacy</td>\n",
              "      <td>True</td>\n",
              "      <td>\"However, UGT2B7 rs73823859 showed a statistic...</td>\n",
              "      <td>UGT2B7</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5264</th>\n",
              "      <td>1448531887</td>\n",
              "      <td>rs28365063</td>\n",
              "      <td>PMID:28063968</td>\n",
              "      <td>morphine</td>\n",
              "      <td>Allele G is not associated with exposure to mo...</td>\n",
              "      <td>no</td>\n",
              "      <td>= 0.51</td>\n",
              "      <td>22.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>European</td>\n",
              "      <td>Metabolism/PK</td>\n",
              "      <td>False</td>\n",
              "      <td>Measures assessed were AUC, transit compartmen...</td>\n",
              "      <td>UGT2B7</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5265</th>\n",
              "      <td>1448531882</td>\n",
              "      <td>rs7438135</td>\n",
              "      <td>PMID:28063968</td>\n",
              "      <td>morphine</td>\n",
              "      <td>Allele G is not associated with exposure to mo...</td>\n",
              "      <td>no</td>\n",
              "      <td>= 0.32</td>\n",
              "      <td>22.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>European</td>\n",
              "      <td>Metabolism/PK</td>\n",
              "      <td>False</td>\n",
              "      <td>Measures assessed were AUC, transit compartmen...</td>\n",
              "      <td>UGT2B7</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>5266 rows Ã— 17 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-2a6a9d61-1260-40a6-86a3-6b830165b439')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-2a6a9d61-1260-40a6-86a3-6b830165b439 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-2a6a9d61-1260-40a6-86a3-6b830165b439');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 93
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Arthitecture and Principal:\n",
        "This code should be able to rearrange data downloaded from PharmGKB with only two added columns (\"Alter\" and \"Ref\"), and produces a usable datasheet with chormosome, location, alter, ref and a label of pharmacogenetic significance."
      ],
      "metadata": {
        "id": "GyzeTCsFnJa6"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Load data\n",
        "#data = pd.read_csv(\"/content/gdrive/My Drive/Colab Notebooks/ADME Project/listed_core_o.csv\")"
      ],
      "metadata": {
        "id": "kJ0XMKIPpHht"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Only select SNPs which pharmacogenetic significance is known\n",
        "\n",
        "data['numeric'] = data['Significance'].map({'yes': 1, 'no': 0})\n",
        "data_dropped = data.dropna(subset=['numeric','Alter'])"
      ],
      "metadata": {
        "id": "fVv4CRnGp9MI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Sort with variant, alternative base and significant label\n",
        "#This sorted data set can refered later for certain variant location\n",
        "#as it includes all related literature.\n",
        "\n",
        "data_sorted = data_dropped.sort_values(['Variant','Alter','Significance'])"
      ],
      "metadata": {
        "id": "sHq7F-QMx0Gx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# If for the same alternative base and same variant location, there are both significant report and non-siginificant report,\n",
        "#delete the later.\n",
        "\n",
        "data_unique = data_sorted.drop_duplicates(subset=['Variant', 'Alter'], keep='last')"
      ],
      "metadata": {
        "id": "vBq2eLE40DSW"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "import requests\n",
        "\n",
        "# Create empty lists to store the genomic information\n",
        "chromosomes = []\n",
        "locations = []\n",
        "reference_bases = []\n",
        "\n",
        "# Base URL for Ensembl REST API\n",
        "base_url = 'https://rest.ensembl.org'\n",
        "\n",
        "# Iterate over rsIDs\n",
        "for rsID in data_unique['Variant']:\n",
        "    # Construct the API endpoint URL\n",
        "    endpoint = f'{base_url}/variation/human/{rsID}?phenotypes=1'\n",
        "\n",
        "    # Send GET request to the API\n",
        "    response = requests.get(endpoint, headers={'Content-Type': 'application/json'})\n",
        "\n",
        "    # Check if the request was successful\n",
        "    if response.status_code == 200:\n",
        "        # Extract relevant information from the response\n",
        "        data = response.json()\n",
        "        chromosome = data['mappings'][0]['seq_region_name']\n",
        "        location = data['mappings'][0]['start']\n",
        "        reference_base = data['mappings'][0]['allele_string'].split('/')[0]\n",
        "\n",
        "        # Append the genomic information to the respective lists\n",
        "        chromosomes.append(chromosome)\n",
        "        locations.append(location)\n",
        "        reference_bases.append(reference_base)\n",
        "    else:\n",
        "        # If there was an error, append None to the lists\n",
        "        chromosomes.append(None)\n",
        "        locations.append(None)\n",
        "        reference_bases.append(None)\n",
        "\n",
        "# Create a new DataFrame with the genomic information\n",
        "search_df = pd.DataFrame({\n",
        "    'Chromosome': chromosomes,\n",
        "    'Location': locations,\n",
        "    'Reference Base': reference_bases\n",
        "})"
      ],
      "metadata": {
        "id": "uRDcicXhaIQz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Combine the search results with orginal dataset\n",
        "search_df.set_index(data_unique.index, inplace=True)\n",
        "combined_df = data_unique.join(search_df)\n",
        "#combined_df.to_csv('/content/gdrive/My Drive/Colab Notebooks/ADME Project/Unlisted_auto_v1.csv')"
      ],
      "metadata": {
        "id": "Mck5jynofKV_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "markdown",
      "source": [
        "*Trim the code with GRCh38 setting.\n",
        "error soruce: https://rest.ensembl.org/vep/human/region/22:24034135:24034135/A\n",
        "Checked Unlisted_o, further debugging:\n",
        "https://rest.ensembl.org/variation/human/rs4630?content-type=application/json"
      ],
      "metadata": {
        "id": "fwVtJegqqr22"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Filter duplicates\n",
        "\n",
        "combined_df['Combination'] = combined_df[['Alter', 'Ref']].apply(lambda x: ''.join(sorted(x.values.astype(str))), axis=1)\n",
        "# Drop rows where Column1 and Combination values are the same, and Column4 is 'no'\n",
        "filterd_df = combined_df.dropna(subset=['Chromosome','Location','Reference Base'])\n",
        "filterd_df = filterd_df[~((combined_df.duplicated(subset=['Location', 'Combination']) & (combined_df['numeric'] == 0)) |\n",
        "      ((combined_df['numeric'] == 1) & combined_df.duplicated(subset=['Location', 'Combination']) & (combined_df.groupby('Location')['numeric'].transform('max') == 1)))]"
      ],
      "metadata": {
        "id": "p--1R7DmJ-nK"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Finalizing\n",
        "#\"Pediatric\" deleted due to irrelavance\n",
        "final1 = filterd_df[['Chromosome','Genes','Location','Reference Base','Alter','Combination','numeric','Literature','Drugs','Association','More Details']]"
      ],
      "metadata": {
        "id": "ZM54ujM4Z-Qz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Filter again\n",
        "#1)Combination Combination - Reference Base\n",
        "#2)Delete nan Combination\n",
        "#3)Change the rest Combination to \"True Alter\"\n",
        "filtered2 = final1\n",
        "filtered2['Combination'] = final1.apply(lambda row: str(row['Combination']).replace(str(row['Reference Base']), ''), axis=1)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ISlFuN_RcZsH",
        "outputId": "bf75c406-aa73-418a-b7e8-a560713c4632"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-101-4d14096c7fa8>:6: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame.\n",
            "Try using .loc[row_indexer,col_indexer] = value instead\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  filtered2['Combination'] = final1.apply(lambda row: str(row['Combination']).replace(str(row['Reference Base']), ''), axis=1)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "data_dropped2 = filtered2[~(filtered2['Combination'] == 'nan')]"
      ],
      "metadata": {
        "id": "t3Ba6V6DJxci"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Final 2: Calculate True Alter\n",
        "final2 = data_dropped2\n",
        "final2['Combination'] = data_dropped2.apply(lambda row: row['Alter'] if str(row['Alter']) != str(row['Reference Base']) else str(row['Combination']), axis=1)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9IrJzXp-M3pD",
        "outputId": "5d024abf-25d0-4a69-8e3e-87891da1d0cb"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-103-80a9026e924b>:3: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame.\n",
            "Try using .loc[row_indexer,col_indexer] = value instead\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  final2['Combination'] = data_dropped2.apply(lambda row: row['Alter'] if str(row['Alter']) != str(row['Reference Base']) else str(row['Combination']), axis=1)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Final 3: Delete irrelevant column\n",
        "final3 = final2.drop(columns='Alter')\n",
        "final3 = final3.rename(columns={'Combination': 'Altered Base'})\n",
        "final3 = final3[~(final3['Altered Base'] == None)]\n",
        "final3"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 658
        },
        "id": "yCR-uRtTN71V",
        "outputId": "e2f5730d-1468-449e-9ab1-196945e5e5a7"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "     Chromosome   Genes   Location Reference Base Altered Base  numeric  \\\n",
              "5148          4  UGT2B7   69112090              T            C      0.0   \n",
              "1603          4   ABCG2   88169725              T            C      0.0   \n",
              "5108          4  UGT2B7   69105219              A            C      0.0   \n",
              "4231         22   GSTT1   23898914              C            T      1.0   \n",
              "973           7   ABCB1   87535576              T            A      0.0   \n",
              "...         ...     ...        ...            ...          ...      ...   \n",
              "1719         19  CYP2A6   40848591              C            A      0.0   \n",
              "4791          2  UGT1A1  233772999              G            C      1.0   \n",
              "649           7   ABCB1   87560614              T            G      0.0   \n",
              "4813          2  UGT1A1  233759924              C            T      0.0   \n",
              "587           7   ABCB1   87600124              T            C      1.0   \n",
              "\n",
              "            Literature                  Drugs  \\\n",
              "5148     PMID:30920410               morphine   \n",
              "1603     PMID:30924126            allopurinol   \n",
              "5108  PMCID:PMC5521342               fentanyl   \n",
              "4231  PMCID:PMC3564008                    NaN   \n",
              "973      PMID:26421491          carbamazepine   \n",
              "...                ...                    ...   \n",
              "1719     PMID:23172109              efavirenz   \n",
              "4791     PMID:25151207  atazanavir; ritonavir   \n",
              "649      PMID:30713339               imatinib   \n",
              "4813  PMCID:PMC5508045               warfarin   \n",
              "587   PMCID:PMC5354739           methotrexate   \n",
              "\n",
              "                                            Association  \\\n",
              "5148  Allele C is not associated with metabolism of ...   \n",
              "1603  Allele T is associated with decreased response...   \n",
              "5108  Allele A is not associated with dose of fentan...   \n",
              "4231  Allele T is associated with increased transcri...   \n",
              "973   Allele A is not associated with concentrations...   \n",
              "...                                                 ...   \n",
              "1719  Allele A is not associated with metabolism of ...   \n",
              "4791  Allele G is associated with increased risk of ...   \n",
              "649   Allele T is not associated with risk of Drug T...   \n",
              "4813  Allele C is not associated with dose of warfar...   \n",
              "587   Allele C is associated with increased concentr...   \n",
              "\n",
              "                                           More Details  \n",
              "5148  Variant was potentially associated with morphi...  \n",
              "1603  Response to allopurinol was determined by whet...  \n",
              "5108  There was no significant difference in 24 hour...  \n",
              "4231                                                NaN  \n",
              "973                                                 NaN  \n",
              "...                                                 ...  \n",
              "1719  No association was seen between this polymorph...  \n",
              "4791  The authors compared the frequencies of the G ...  \n",
              "649   No significant association between genotype an...  \n",
              "4813                                                NaN  \n",
              "587   Concentrations refers to area under the concen...  \n",
              "\n",
              "[353 rows x 10 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-30d93d5f-6e3f-4e8f-ad51-7e66b898bdf7\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Chromosome</th>\n",
              "      <th>Genes</th>\n",
              "      <th>Location</th>\n",
              "      <th>Reference Base</th>\n",
              "      <th>Altered Base</th>\n",
              "      <th>numeric</th>\n",
              "      <th>Literature</th>\n",
              "      <th>Drugs</th>\n",
              "      <th>Association</th>\n",
              "      <th>More Details</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>5148</th>\n",
              "      <td>4</td>\n",
              "      <td>UGT2B7</td>\n",
              "      <td>69112090</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMID:30920410</td>\n",
              "      <td>morphine</td>\n",
              "      <td>Allele C is not associated with metabolism of ...</td>\n",
              "      <td>Variant was potentially associated with morphi...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1603</th>\n",
              "      <td>4</td>\n",
              "      <td>ABCG2</td>\n",
              "      <td>88169725</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMID:30924126</td>\n",
              "      <td>allopurinol</td>\n",
              "      <td>Allele T is associated with decreased response...</td>\n",
              "      <td>Response to allopurinol was determined by whet...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5108</th>\n",
              "      <td>4</td>\n",
              "      <td>UGT2B7</td>\n",
              "      <td>69105219</td>\n",
              "      <td>A</td>\n",
              "      <td>C</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMCID:PMC5521342</td>\n",
              "      <td>fentanyl</td>\n",
              "      <td>Allele A is not associated with dose of fentan...</td>\n",
              "      <td>There was no significant difference in 24 hour...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4231</th>\n",
              "      <td>22</td>\n",
              "      <td>GSTT1</td>\n",
              "      <td>23898914</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMCID:PMC3564008</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Allele T is associated with increased transcri...</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>973</th>\n",
              "      <td>7</td>\n",
              "      <td>ABCB1</td>\n",
              "      <td>87535576</td>\n",
              "      <td>T</td>\n",
              "      <td>A</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMID:26421491</td>\n",
              "      <td>carbamazepine</td>\n",
              "      <td>Allele A is not associated with concentrations...</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1719</th>\n",
              "      <td>19</td>\n",
              "      <td>CYP2A6</td>\n",
              "      <td>40848591</td>\n",
              "      <td>C</td>\n",
              "      <td>A</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMID:23172109</td>\n",
              "      <td>efavirenz</td>\n",
              "      <td>Allele A is not associated with metabolism of ...</td>\n",
              "      <td>No association was seen between this polymorph...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4791</th>\n",
              "      <td>2</td>\n",
              "      <td>UGT1A1</td>\n",
              "      <td>233772999</td>\n",
              "      <td>G</td>\n",
              "      <td>C</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMID:25151207</td>\n",
              "      <td>atazanavir; ritonavir</td>\n",
              "      <td>Allele G is associated with increased risk of ...</td>\n",
              "      <td>The authors compared the frequencies of the G ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>649</th>\n",
              "      <td>7</td>\n",
              "      <td>ABCB1</td>\n",
              "      <td>87560614</td>\n",
              "      <td>T</td>\n",
              "      <td>G</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMID:30713339</td>\n",
              "      <td>imatinib</td>\n",
              "      <td>Allele T is not associated with risk of Drug T...</td>\n",
              "      <td>No significant association between genotype an...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4813</th>\n",
              "      <td>2</td>\n",
              "      <td>UGT1A1</td>\n",
              "      <td>233759924</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMCID:PMC5508045</td>\n",
              "      <td>warfarin</td>\n",
              "      <td>Allele C is not associated with dose of warfar...</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>587</th>\n",
              "      <td>7</td>\n",
              "      <td>ABCB1</td>\n",
              "      <td>87600124</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMCID:PMC5354739</td>\n",
              "      <td>methotrexate</td>\n",
              "      <td>Allele C is associated with increased concentr...</td>\n",
              "      <td>Concentrations refers to area under the concen...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>353 rows Ã— 10 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-30d93d5f-6e3f-4e8f-ad51-7e66b898bdf7')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-30d93d5f-6e3f-4e8f-ad51-7e66b898bdf7 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-30d93d5f-6e3f-4e8f-ad51-7e66b898bdf7');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 104
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "np.shape(final3)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FdQxQ11hUqEm",
        "outputId": "6446f395-005b-4d4d-b015-3510abc828f4"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(353, 10)"
            ]
          },
          "metadata": {},
          "execution_count": 105
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "final4 = final3.sort_values(['Chromosome', 'Genes'])\n",
        "final4"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 658
        },
        "id": "FMbPyF1CjpKN",
        "outputId": "c2e0d45c-2b98-4158-f658-5fc3b57712f1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "              Chromosome  Genes  Location Reference Base Altered Base  \\\n",
              "3622                   1   DPYD  97549600              T            C   \n",
              "3663                   1   DPYD  97305279              G            A   \n",
              "3752                   1   DPYD  97078987              G            T   \n",
              "3563                   1   DPYD  97699474              T            C   \n",
              "3974                   1   DPYD  97593438              A            C   \n",
              "...                  ...    ...       ...            ...          ...   \n",
              "4452                   8   NAT2  18390758              G            C   \n",
              "4301                   8   NAT2  18400616              A            T   \n",
              "4404                   8   NAT2  18390208              T            A   \n",
              "4327                   8   NAT2  18390403              A            G   \n",
              "4217  CHR_HSCHR22_1_CTG7  GSTT1  24034135              G            A   \n",
              "\n",
              "      numeric        Literature  \\\n",
              "3622      1.0  PMCID:PMC4012613   \n",
              "3663      1.0  PMCID:PMC4012613   \n",
              "3752      1.0  PMCID:PMC5309195   \n",
              "3563      1.0  PMCID:PMC4012613   \n",
              "3974      1.0     PMID:19530960   \n",
              "...       ...               ...   \n",
              "4452      1.0     PMID:20602614   \n",
              "4301      0.0  PMCID:PMC3619396   \n",
              "4404      1.0     PMID:19891553   \n",
              "4327      0.0     PMID:19891553   \n",
              "4217      0.0     PMID:19504558   \n",
              "\n",
              "                                                  Drugs  \\\n",
              "3622                                       fluorouracil   \n",
              "3663                                       fluorouracil   \n",
              "3752                                                NaN   \n",
              "3563                                       fluorouracil   \n",
              "3974                         capecitabine; fluorouracil   \n",
              "...                                                 ...   \n",
              "4452                                            aspirin   \n",
              "4301                    sulfamethoxazole / trimethoprim   \n",
              "4404      ethambutol; isoniazid; pyrazinamide; rifampin   \n",
              "4327      ethambutol; isoniazid; pyrazinamide; rifampin   \n",
              "4217  dexamethasone; diphenhydramine; paclitaxel; ra...   \n",
              "\n",
              "                                            Association  \\\n",
              "3622  Allele C is associated with decreased activity...   \n",
              "3663  Allele A is associated with decreased activity...   \n",
              "3752  Allele T is associated with decreased catalyti...   \n",
              "3563  Allele C is associated with decreased activity...   \n",
              "3974  Allele C is associated with increased severity...   \n",
              "...                                                 ...   \n",
              "4452  Allele C is associated with increased risk of ...   \n",
              "4301  Allele T is not associated with increased risk...   \n",
              "4404  Allele A is associated with increased risk of ...   \n",
              "4327  Allele G is not associated with risk of Hepati...   \n",
              "4217  Allele G is not associated with overall surviv...   \n",
              "\n",
              "                                           More Details  \n",
              "3622  Functional study. The C allele was associated ...  \n",
              "3663  Functional study. The A allele was associated ...  \n",
              "3752  This variant is also called p.P1023S and is fo...  \n",
              "3563  Functional study. The C allele was associated ...  \n",
              "3974  The C allele was more common in those with Gra...  \n",
              "...                                                 ...  \n",
              "4452  The C allele was found at a higher frequency i...  \n",
              "4301  Minor allele frequencies were compared between...  \n",
              "4404  The frequency of the A allele was significantl...  \n",
              "4327  The frequency of the G allele or genotype freq...  \n",
              "4217  Although not significant alone, when looked at...  \n",
              "\n",
              "[353 rows x 10 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-0b467c30-b363-4d95-96b9-3d2958ce1d2d\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Chromosome</th>\n",
              "      <th>Genes</th>\n",
              "      <th>Location</th>\n",
              "      <th>Reference Base</th>\n",
              "      <th>Altered Base</th>\n",
              "      <th>numeric</th>\n",
              "      <th>Literature</th>\n",
              "      <th>Drugs</th>\n",
              "      <th>Association</th>\n",
              "      <th>More Details</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>3622</th>\n",
              "      <td>1</td>\n",
              "      <td>DPYD</td>\n",
              "      <td>97549600</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMCID:PMC4012613</td>\n",
              "      <td>fluorouracil</td>\n",
              "      <td>Allele C is associated with decreased activity...</td>\n",
              "      <td>Functional study. The C allele was associated ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3663</th>\n",
              "      <td>1</td>\n",
              "      <td>DPYD</td>\n",
              "      <td>97305279</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMCID:PMC4012613</td>\n",
              "      <td>fluorouracil</td>\n",
              "      <td>Allele A is associated with decreased activity...</td>\n",
              "      <td>Functional study. The A allele was associated ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3752</th>\n",
              "      <td>1</td>\n",
              "      <td>DPYD</td>\n",
              "      <td>97078987</td>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMCID:PMC5309195</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Allele T is associated with decreased catalyti...</td>\n",
              "      <td>This variant is also called p.P1023S and is fo...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3563</th>\n",
              "      <td>1</td>\n",
              "      <td>DPYD</td>\n",
              "      <td>97699474</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMCID:PMC4012613</td>\n",
              "      <td>fluorouracil</td>\n",
              "      <td>Allele C is associated with decreased activity...</td>\n",
              "      <td>Functional study. The C allele was associated ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3974</th>\n",
              "      <td>1</td>\n",
              "      <td>DPYD</td>\n",
              "      <td>97593438</td>\n",
              "      <td>A</td>\n",
              "      <td>C</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMID:19530960</td>\n",
              "      <td>capecitabine; fluorouracil</td>\n",
              "      <td>Allele C is associated with increased severity...</td>\n",
              "      <td>The C allele was more common in those with Gra...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4452</th>\n",
              "      <td>8</td>\n",
              "      <td>NAT2</td>\n",
              "      <td>18390758</td>\n",
              "      <td>G</td>\n",
              "      <td>C</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMID:20602614</td>\n",
              "      <td>aspirin</td>\n",
              "      <td>Allele C is associated with increased risk of ...</td>\n",
              "      <td>The C allele was found at a higher frequency i...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4301</th>\n",
              "      <td>8</td>\n",
              "      <td>NAT2</td>\n",
              "      <td>18400616</td>\n",
              "      <td>A</td>\n",
              "      <td>T</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMCID:PMC3619396</td>\n",
              "      <td>sulfamethoxazole / trimethoprim</td>\n",
              "      <td>Allele T is not associated with increased risk...</td>\n",
              "      <td>Minor allele frequencies were compared between...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4404</th>\n",
              "      <td>8</td>\n",
              "      <td>NAT2</td>\n",
              "      <td>18390208</td>\n",
              "      <td>T</td>\n",
              "      <td>A</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMID:19891553</td>\n",
              "      <td>ethambutol; isoniazid; pyrazinamide; rifampin</td>\n",
              "      <td>Allele A is associated with increased risk of ...</td>\n",
              "      <td>The frequency of the A allele was significantl...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4327</th>\n",
              "      <td>8</td>\n",
              "      <td>NAT2</td>\n",
              "      <td>18390403</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMID:19891553</td>\n",
              "      <td>ethambutol; isoniazid; pyrazinamide; rifampin</td>\n",
              "      <td>Allele G is not associated with risk of Hepati...</td>\n",
              "      <td>The frequency of the G allele or genotype freq...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4217</th>\n",
              "      <td>CHR_HSCHR22_1_CTG7</td>\n",
              "      <td>GSTT1</td>\n",
              "      <td>24034135</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMID:19504558</td>\n",
              "      <td>dexamethasone; diphenhydramine; paclitaxel; ra...</td>\n",
              "      <td>Allele G is not associated with overall surviv...</td>\n",
              "      <td>Although not significant alone, when looked at...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>353 rows Ã— 10 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-0b467c30-b363-4d95-96b9-3d2958ce1d2d')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-0b467c30-b363-4d95-96b9-3d2958ce1d2d button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-0b467c30-b363-4d95-96b9-3d2958ce1d2d');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 106
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "final4.to_csv('/content/gdrive/My Drive/Colab Notebooks/ADME Project/Unlisted_auto.csv')"
      ],
      "metadata": {
        "id": "Q8SY8WmRivqO"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}