{
  "nbformat": 4,
  "nbformat_minor": 0,
  "metadata": {
    "colab": {
      "provenance": [],
      "authorship_tag": "ABX9TyPI+ZLaIAZINCr+9Z8P+xdK",
      "include_colab_link": true
    },
    "kernelspec": {
      "name": "python3",
      "display_name": "Python 3"
    },
    "language_info": {
      "name": "python"
    }
  },
  "cells": [
    {
      "cell_type": "markdown",
      "metadata": {
        "id": "view-in-github",
        "colab_type": "text"
      },
      "source": [
        "<a href=\"https://colab.research.google.com/github/katha815/ADME-Classifier/blob/main/Dataset_Expansion6_autoListedCore.ipynb\" target=\"_parent\"><img src=\"https://colab.research.google.com/assets/colab-badge.svg\" alt=\"Open In Colab\"/></a>"
      ]
    },
    {
      "cell_type": "code",
      "execution_count": null,
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "m_9HrF20mF_F",
        "outputId": "f4ef6b53-0a82-4c4f-c3e3-9a7729409a9d"
      },
      "outputs": [
        {
          "output_type": "stream",
          "name": "stdout",
          "text": [
            "Mounted at /content/gdrive\n"
          ]
        }
      ],
      "source": [
        "from google.colab import drive\n",
        "drive.mount('/content/gdrive')\n",
        "import numpy as np\n",
        "import pandas as pd\n",
        "import matplotlib.pyplot as plt"
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "Results are reliable: same data size with manual version!"
      ],
      "metadata": {
        "id": "3LNeQmUmjQtp"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "Core_genes = pd.read_csv(\"/content/gdrive/My Drive/Colab Notebooks/ADME Project/Core ADME Genes.csv\")"
      ],
      "metadata": {
        "id": "a-W2aKLLmeo1"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Data preparation (real run)\n",
        "file_paths = []\n",
        "\n",
        "# List of TSV file paths\n",
        "for i in range(len(Core_genes['Gene'])):\n",
        "  if Core_genes['Listed?'][i]==1:\n",
        "    gene_name = Core_genes['Gene'][i]\n",
        "    file_paths.append(f'/content/gdrive/My Drive/Colab Notebooks/ADME Project/{gene_name}.tsv')\n",
        "\n",
        "# Create an empty list to store DataFrames\n",
        "dfs = []\n",
        "\n",
        "# Read each TSV file and append its DataFrame to the list\n",
        "for file_path in file_paths:\n",
        "    try:\n",
        "        df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
        "        dfs.append(df)\n",
        "    except pd.errors.ParserError:\n",
        "        print(f\"Error reading file: {file_path}\")\n",
        "\n",
        "# Concatenate the DataFrames into a single DataFrame\n",
        "raw_df = pd.concat(dfs, ignore_index=True)\n",
        "\n",
        "# Save the combined DataFrame as a CSV file\n",
        "#raw_df.to_csv('/content/gdrive/My Drive/Colab Notebooks/ADME Project/ListedCore_raw_auto.csv', index=False)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "q_9EKEWQjgTL",
        "outputId": "40656c41-a8eb-440e-ba00-a0278ecf462e"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-3-1b968528f97f>:16: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-3-1b968528f97f>:16: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-3-1b968528f97f>:16: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-3-1b968528f97f>:16: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-3-1b968528f97f>:16: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-3-1b968528f97f>:16: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-3-1b968528f97f>:16: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-3-1b968528f97f>:16: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-3-1b968528f97f>:16: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-3-1b968528f97f>:16: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-3-1b968528f97f>:16: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-3-1b968528f97f>:16: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n",
            "<ipython-input-3-1b968528f97f>:16: FutureWarning: The error_bad_lines argument has been deprecated and will be removed in a future version. Use on_bad_lines in the future.\n",
            "\n",
            "\n",
            "  df = pd.read_csv(file_path, sep='\\t', error_bad_lines=False)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Ref/Alter rows creation (Real-run)\n",
        "data = raw_df\n",
        "\n",
        "# Define the search patterns and corresponding replacement values\n",
        "patterns = ['Allele A ', 'Allele C ', 'Allele G ', 'Allele T ']\n",
        "patterns2 = ['allele A', 'allele C', 'allele G', 'allele T']\n",
        "replacements = ['A', 'C', 'G', 'T']\n",
        "\n",
        "# Create the two new columns using the search patterns and replacements\n",
        "data['Alter'] = data['Association'].apply(lambda x: next((replacements[i] for i, pattern in enumerate(patterns) if pattern in x), np.nan))\n",
        "data['Ref'] = data['Association'].apply(lambda x: next((replacements[i] for i, pattern2 in enumerate(patterns2) if pattern2 in x), np.nan))\n",
        "\n",
        "data"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "id": "80CYdt8-r3OL",
        "outputId": "6a324081-707e-40d0-8f85-056ed4b39de1"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "      PharmGKB ID                           Variant        Literature  \\\n",
              "0       827921753                         rs3740066     PMID:22630058   \n",
              "1       981939846                         rs3740066     PMID:18444945   \n",
              "2      1451129980                          rs717620     PMID:31648372   \n",
              "3       827824467                         rs3740066     PMID:22256867   \n",
              "4       827844958                         rs2804398     PMID:21449672   \n",
              "...           ...                               ...               ...   \n",
              "6552   1451893624                         rs1142345     PMID:36143183   \n",
              "6553   1184349029  TPMT*1; TPMT*2; TPMT*3A; TPMT*3C  PMCID:PMC3080646   \n",
              "6554   1000539737                        rs12201199  PMCID:PMC3883563   \n",
              "6555   1184169832                   TPMT*1; TPMT*3A     PMID:18987660   \n",
              "6556   1000539742                         rs1800460  PMCID:PMC3883563   \n",
              "\n",
              "                                    Drugs  \\\n",
              "0                          antiepileptics   \n",
              "1     cyclosporine; mycophenolate mofetil   \n",
              "2     Drugs For Treatment Of Tuberculosis   \n",
              "3                          antiepileptics   \n",
              "4                          antiepileptics   \n",
              "...                                   ...   \n",
              "6552                         azathioprine   \n",
              "6553                       mercaptopurine   \n",
              "6554                            cisplatin   \n",
              "6555                       mercaptopurine   \n",
              "6556                            cisplatin   \n",
              "\n",
              "                                            Association Significance  \\\n",
              "0     Genotypes CT + TT are associated with increase...          yes   \n",
              "1     Genotypes CT + TT are not associated with incr...           no   \n",
              "2     Allele T is not associated with risk of Toxic ...           no   \n",
              "3     Genotypes CT + TT are not associated with incr...           no   \n",
              "4     Allele T is not associated with resistance to ...           no   \n",
              "...                                                 ...          ...   \n",
              "6552  Genotypes CC + CT (assigned as poor metabolize...          yes   \n",
              "6553  TPMT *1/*2 + *1/*3A + *1/*3C is associated wit...          yes   \n",
              "6554  Allele T is not associated with risk of Ototox...           no   \n",
              "6555  TPMT *1/*3A is associated with decreased dose ...          yes   \n",
              "6556  Allele T is not associated with risk of Ototox...           no   \n",
              "\n",
              "       P-Value  # of Cases  # of Controls Biogeographical Groups  \\\n",
              "0      = 0.038       217.0          320.0             East Asian   \n",
              "1       = 0.63       237.0            NaN               European   \n",
              "2      = 0.514       746.0            NaN             East Asian   \n",
              "3        = 0.8       357.0          247.0               European   \n",
              "4       = 0.75       586.0            NaN             East Asian   \n",
              "...        ...         ...            ...                    ...   \n",
              "6552  = 0.0004      1050.0            NaN             East Asian   \n",
              "6553  < 0.0001        65.0            NaN                Unknown   \n",
              "6554     = 0.5        64.0          149.0        Multiple groups   \n",
              "6555   < 0.001       312.0            NaN               European   \n",
              "6556    = 0.11        64.0          149.0        Multiple groups   \n",
              "\n",
              "     Phenotype Categories  Pediatric  \\\n",
              "0                Efficacy      False   \n",
              "1                Efficacy      False   \n",
              "2                Toxicity      False   \n",
              "3                Efficacy      False   \n",
              "4                Efficacy      False   \n",
              "...                   ...        ...   \n",
              "6552             Toxicity      False   \n",
              "6553        Metabolism/PK       True   \n",
              "6554             Toxicity       True   \n",
              "6555               Dosage       True   \n",
              "6556             Toxicity       True   \n",
              "\n",
              "                                           More Details  Genes Alter  Ref  \n",
              "0     This SNP was also part of haplotype associated...  ABCC2   NaN  NaN  \n",
              "1     No association was found with biopsy-proven ac...  ABCC2   NaN  NaN  \n",
              "2     Please note that alleles have been complemente...  ABCC2     T    C  \n",
              "3                                                   NaN  ABCC2   NaN  NaN  \n",
              "4     Epilepsy patients on antiepileptic drug (AED) ...  ABCC2     T    A  \n",
              "...                                                 ...    ...   ...  ...  \n",
              "6552  Effect is described for \"non-NM genotype\", \"A ...   TPMT   NaN  NaN  \n",
              "6553  ITPA SNPs were assessed as well and included i...   TPMT   NaN  NaN  \n",
              "6554                                                NaN   TPMT     T    A  \n",
              "6555  Patients were in the maintenance phase of trea...   TPMT   NaN  NaN  \n",
              "6556                                                NaN   TPMT     T    C  \n",
              "\n",
              "[6557 rows x 16 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-7517fa3b-d924-4d59-84b5-eb909d457b7a\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>PharmGKB ID</th>\n",
              "      <th>Variant</th>\n",
              "      <th>Literature</th>\n",
              "      <th>Drugs</th>\n",
              "      <th>Association</th>\n",
              "      <th>Significance</th>\n",
              "      <th>P-Value</th>\n",
              "      <th># of Cases</th>\n",
              "      <th># of Controls</th>\n",
              "      <th>Biogeographical Groups</th>\n",
              "      <th>Phenotype Categories</th>\n",
              "      <th>Pediatric</th>\n",
              "      <th>More Details</th>\n",
              "      <th>Genes</th>\n",
              "      <th>Alter</th>\n",
              "      <th>Ref</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>0</th>\n",
              "      <td>827921753</td>\n",
              "      <td>rs3740066</td>\n",
              "      <td>PMID:22630058</td>\n",
              "      <td>antiepileptics</td>\n",
              "      <td>Genotypes CT + TT are associated with increase...</td>\n",
              "      <td>yes</td>\n",
              "      <td>= 0.038</td>\n",
              "      <td>217.0</td>\n",
              "      <td>320.0</td>\n",
              "      <td>East Asian</td>\n",
              "      <td>Efficacy</td>\n",
              "      <td>False</td>\n",
              "      <td>This SNP was also part of haplotype associated...</td>\n",
              "      <td>ABCC2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>1</th>\n",
              "      <td>981939846</td>\n",
              "      <td>rs3740066</td>\n",
              "      <td>PMID:18444945</td>\n",
              "      <td>cyclosporine; mycophenolate mofetil</td>\n",
              "      <td>Genotypes CT + TT are not associated with incr...</td>\n",
              "      <td>no</td>\n",
              "      <td>= 0.63</td>\n",
              "      <td>237.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>European</td>\n",
              "      <td>Efficacy</td>\n",
              "      <td>False</td>\n",
              "      <td>No association was found with biopsy-proven ac...</td>\n",
              "      <td>ABCC2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2</th>\n",
              "      <td>1451129980</td>\n",
              "      <td>rs717620</td>\n",
              "      <td>PMID:31648372</td>\n",
              "      <td>Drugs For Treatment Of Tuberculosis</td>\n",
              "      <td>Allele T is not associated with risk of Toxic ...</td>\n",
              "      <td>no</td>\n",
              "      <td>= 0.514</td>\n",
              "      <td>746.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>East Asian</td>\n",
              "      <td>Toxicity</td>\n",
              "      <td>False</td>\n",
              "      <td>Please note that alleles have been complemente...</td>\n",
              "      <td>ABCC2</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3</th>\n",
              "      <td>827824467</td>\n",
              "      <td>rs3740066</td>\n",
              "      <td>PMID:22256867</td>\n",
              "      <td>antiepileptics</td>\n",
              "      <td>Genotypes CT + TT are not associated with incr...</td>\n",
              "      <td>no</td>\n",
              "      <td>= 0.8</td>\n",
              "      <td>357.0</td>\n",
              "      <td>247.0</td>\n",
              "      <td>European</td>\n",
              "      <td>Efficacy</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>ABCC2</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>4</th>\n",
              "      <td>827844958</td>\n",
              "      <td>rs2804398</td>\n",
              "      <td>PMID:21449672</td>\n",
              "      <td>antiepileptics</td>\n",
              "      <td>Allele T is not associated with resistance to ...</td>\n",
              "      <td>no</td>\n",
              "      <td>= 0.75</td>\n",
              "      <td>586.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>East Asian</td>\n",
              "      <td>Efficacy</td>\n",
              "      <td>False</td>\n",
              "      <td>Epilepsy patients on antiepileptic drug (AED) ...</td>\n",
              "      <td>ABCC2</td>\n",
              "      <td>T</td>\n",
              "      <td>A</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6552</th>\n",
              "      <td>1451893624</td>\n",
              "      <td>rs1142345</td>\n",
              "      <td>PMID:36143183</td>\n",
              "      <td>azathioprine</td>\n",
              "      <td>Genotypes CC + CT (assigned as poor metabolize...</td>\n",
              "      <td>yes</td>\n",
              "      <td>= 0.0004</td>\n",
              "      <td>1050.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>East Asian</td>\n",
              "      <td>Toxicity</td>\n",
              "      <td>False</td>\n",
              "      <td>Effect is described for \"non-NM genotype\", \"A ...</td>\n",
              "      <td>TPMT</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6553</th>\n",
              "      <td>1184349029</td>\n",
              "      <td>TPMT*1; TPMT*2; TPMT*3A; TPMT*3C</td>\n",
              "      <td>PMCID:PMC3080646</td>\n",
              "      <td>mercaptopurine</td>\n",
              "      <td>TPMT *1/*2 + *1/*3A + *1/*3C is associated wit...</td>\n",
              "      <td>yes</td>\n",
              "      <td>&lt; 0.0001</td>\n",
              "      <td>65.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Unknown</td>\n",
              "      <td>Metabolism/PK</td>\n",
              "      <td>True</td>\n",
              "      <td>ITPA SNPs were assessed as well and included i...</td>\n",
              "      <td>TPMT</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6554</th>\n",
              "      <td>1000539737</td>\n",
              "      <td>rs12201199</td>\n",
              "      <td>PMCID:PMC3883563</td>\n",
              "      <td>cisplatin</td>\n",
              "      <td>Allele T is not associated with risk of Ototox...</td>\n",
              "      <td>no</td>\n",
              "      <td>= 0.5</td>\n",
              "      <td>64.0</td>\n",
              "      <td>149.0</td>\n",
              "      <td>Multiple groups</td>\n",
              "      <td>Toxicity</td>\n",
              "      <td>True</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TPMT</td>\n",
              "      <td>T</td>\n",
              "      <td>A</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6555</th>\n",
              "      <td>1184169832</td>\n",
              "      <td>TPMT*1; TPMT*3A</td>\n",
              "      <td>PMID:18987660</td>\n",
              "      <td>mercaptopurine</td>\n",
              "      <td>TPMT *1/*3A is associated with decreased dose ...</td>\n",
              "      <td>yes</td>\n",
              "      <td>&lt; 0.001</td>\n",
              "      <td>312.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>European</td>\n",
              "      <td>Dosage</td>\n",
              "      <td>True</td>\n",
              "      <td>Patients were in the maintenance phase of trea...</td>\n",
              "      <td>TPMT</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>6556</th>\n",
              "      <td>1000539742</td>\n",
              "      <td>rs1800460</td>\n",
              "      <td>PMCID:PMC3883563</td>\n",
              "      <td>cisplatin</td>\n",
              "      <td>Allele T is not associated with risk of Ototox...</td>\n",
              "      <td>no</td>\n",
              "      <td>= 0.11</td>\n",
              "      <td>64.0</td>\n",
              "      <td>149.0</td>\n",
              "      <td>Multiple groups</td>\n",
              "      <td>Toxicity</td>\n",
              "      <td>True</td>\n",
              "      <td>NaN</td>\n",
              "      <td>TPMT</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>6557 rows Ã— 16 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-7517fa3b-d924-4d59-84b5-eb909d457b7a')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-7517fa3b-d924-4d59-84b5-eb909d457b7a button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-7517fa3b-d924-4d59-84b5-eb909d457b7a');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 4
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "data.dtypes"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "-axu-O3ut_Cn",
        "outputId": "d097c01c-5094-4529-a29b-1e543d12daed"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "PharmGKB ID                 int64\n",
              "Variant                    object\n",
              "Literature                 object\n",
              "Drugs                      object\n",
              "Association                object\n",
              "Significance               object\n",
              "P-Value                    object\n",
              "# of Cases                float64\n",
              "# of Controls             float64\n",
              "Biogeographical Groups     object\n",
              "Phenotype Categories       object\n",
              "Pediatric                    bool\n",
              "More Details               object\n",
              "Genes                      object\n",
              "Alter                      object\n",
              "Ref                        object\n",
              "dtype: object"
            ]
          },
          "metadata": {},
          "execution_count": 5
        }
      ]
    },
    {
      "cell_type": "markdown",
      "source": [
        "## Arthitecture and Principal:\n",
        "This code should be able to rearrange data downloaded from PharmGKB with only two added columns (\"Alter\" and \"Ref\"), and produces a usable datasheet with chormosome, location, alter, ref and a label of pharmacogenetic significance."
      ],
      "metadata": {
        "id": "GyzeTCsFnJa6"
      }
    },
    {
      "cell_type": "code",
      "source": [
        "# Load data\n",
        "#data = pd.read_csv(\"/content/gdrive/My Drive/Colab Notebooks/ADME Project/listed_core_o.csv\")"
      ],
      "metadata": {
        "id": "kJ0XMKIPpHht"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Only select SNPs which pharmacogenetic significance is known\n",
        "\n",
        "data['numeric'] = data['Significance'].map({'yes': 1, 'no': 0})\n",
        "data_dropped = data.dropna(subset=['numeric','Alter'])"
      ],
      "metadata": {
        "id": "fVv4CRnGp9MI"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Sort with variant, alternative base and significant label\n",
        "#This sorted data set can refered later for certain variant location\n",
        "#as it includes all related literature.\n",
        "\n",
        "data_sorted = data_dropped.sort_values(['Variant','Alter','Significance'])\n",
        "\n",
        "data_sorted.to_csv('/content/gdrive/My Drive/Colab Notebooks/ADME Project/ListedCore_o_auto.csv')"
      ],
      "metadata": {
        "id": "sHq7F-QMx0Gx"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# If for the same alternative base and same variant location, there are both significant report and non-siginificant report,\n",
        "#delete the later.\n",
        "\n",
        "data_unique = data_sorted.drop_duplicates(subset=['Variant', 'Alter'], keep='last')\n",
        "\n",
        "data_unique"
      ],
      "metadata": {
        "id": "vBq2eLE40DSW",
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 1000
        },
        "outputId": "01d5502e-9514-41db-bc09-63f1efca2f2d"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "      PharmGKB ID     Variant        Literature  \\\n",
              "817    1448993544   rs1042389  PMCID:PMC3940150   \n",
              "5966   1450045235  rs10444413     PMID:29791011   \n",
              "2884    982033113  rs10509680     PMID:20833655   \n",
              "2754    981859317  rs10509681  PMCID:PMC3660078   \n",
              "2767   1447954627  rs10509681     PMID:24704000   \n",
              "...           ...         ...               ...   \n",
              "3387   1184757013   rs9332127  PMCID:PMC4134280   \n",
              "2939   1447519965   rs9332197  PMCID:PMC3330749   \n",
              "3279   1446534354   rs9332238  PMCID:PMC4573240   \n",
              "3905    827788986   rs9332239     PMID:15284535   \n",
              "3897   1451919700   rs9332241  PMCID:PMC9537548   \n",
              "\n",
              "                                                  Drugs  \\\n",
              "817   efavirenz; non-nucleoside reverse transcriptas...   \n",
              "5966                                       methotrexate   \n",
              "2884                                           warfarin   \n",
              "2754                                         paclitaxel   \n",
              "2767                              docetaxel; paclitaxel   \n",
              "...                                                 ...   \n",
              "3387                                           warfarin   \n",
              "2939                                           warfarin   \n",
              "3279                                           warfarin   \n",
              "3905                                        tolbutamide   \n",
              "3897                                           warfarin   \n",
              "\n",
              "                                            Association Significance  \\\n",
              "817   Allele C is not associated with response to ef...           no   \n",
              "5966  Allele C is not associated with clearance of m...           no   \n",
              "2884  Allele T is associated with decreased dose of ...          yes   \n",
              "2754  Allele C is associated with increased risk of ...          yes   \n",
              "2767  Allele T is not associated with response to do...           no   \n",
              "...                                                 ...          ...   \n",
              "3387  Allele G is not associated with dose of warfar...           no   \n",
              "2939  Allele C is associated with time to achieve st...          yes   \n",
              "3279  Allele G is associated with increased dose of ...          yes   \n",
              "3905  Allele T is associated with decreased catalyti...          yes   \n",
              "3897  Allele T is associated with decreased metaboli...          yes   \n",
              "\n",
              "         P-Value  # of Cases  # of Controls  \\\n",
              "817       = 0.80        91.0            NaN   \n",
              "5966    > 0.0001       187.0            NaN   \n",
              "2884   = 3.84E-7      1508.0            NaN   \n",
              "2754     = 0.004       411.0            NaN   \n",
              "2767     = 0.888        22.0           36.0   \n",
              "...          ...         ...            ...   \n",
              "3387      > 0.20       787.0            NaN   \n",
              "2939  = 3.45e-04       311.0            NaN   \n",
              "3279   = 4.4e-13       367.0            NaN   \n",
              "3905         NaN         NaN            NaN   \n",
              "3897  = 5.38E-10       524.0            NaN   \n",
              "\n",
              "                                 Biogeographical Groups Phenotype Categories  \\\n",
              "817                                     Multiple groups             Efficacy   \n",
              "5966                                            Unknown        Metabolism/PK   \n",
              "2884                                         East Asian               Dosage   \n",
              "2754  Multiple groups, combined European American an...             Toxicity   \n",
              "2767                                Central/South Asian             Efficacy   \n",
              "...                                                 ...                  ...   \n",
              "3387                                         East Asian               Dosage   \n",
              "2939                         Multiple groups, 98% white             Efficacy   \n",
              "3279                        Multiple groups, Brazilian                Dosage   \n",
              "3905                                                NaN        Metabolism/PK   \n",
              "3897                                Sub-Saharan African        Metabolism/PK   \n",
              "\n",
              "      Pediatric                                       More Details    Genes  \\\n",
              "817       False                                                NaN   CYP2B6   \n",
              "5966      False                                                NaN  SLCO1B1   \n",
              "2884      False                                                NaN   CYP2C9   \n",
              "2754      False  Risk increases per copy of the C allele (indic...   CYP2C8   \n",
              "2767      False  Patients with complete and partial pathologica...   CYP2C8   \n",
              "...         ...                                                ...      ...   \n",
              "3387      False  Samples, genotypes and INRs from a cohort of 5...   CYP2C9   \n",
              "2939      False                                                NaN   CYP2C9   \n",
              "3279      False  The G allele was strongly associated with high...   CYP2C9   \n",
              "3905      False                                       p value 0.05   CYP2C9   \n",
              "3897      False  measured as decreased S-warfarin/R-warfarin ra...   CYP2C9   \n",
              "\n",
              "     Alter  Ref  numeric  \n",
              "817      C  NaN      0.0  \n",
              "5966     C    T      0.0  \n",
              "2884     T    G      1.0  \n",
              "2754     C    T      1.0  \n",
              "2767     T    C      0.0  \n",
              "...    ...  ...      ...  \n",
              "3387     G    C      0.0  \n",
              "2939     C    T      1.0  \n",
              "3279     G    A      1.0  \n",
              "3905     T  NaN      1.0  \n",
              "3897     T    C      1.0  \n",
              "\n",
              "[278 rows x 17 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-a6d331f4-e225-4c9a-b712-3cbba5babff1\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>PharmGKB ID</th>\n",
              "      <th>Variant</th>\n",
              "      <th>Literature</th>\n",
              "      <th>Drugs</th>\n",
              "      <th>Association</th>\n",
              "      <th>Significance</th>\n",
              "      <th>P-Value</th>\n",
              "      <th># of Cases</th>\n",
              "      <th># of Controls</th>\n",
              "      <th>Biogeographical Groups</th>\n",
              "      <th>Phenotype Categories</th>\n",
              "      <th>Pediatric</th>\n",
              "      <th>More Details</th>\n",
              "      <th>Genes</th>\n",
              "      <th>Alter</th>\n",
              "      <th>Ref</th>\n",
              "      <th>numeric</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>817</th>\n",
              "      <td>1448993544</td>\n",
              "      <td>rs1042389</td>\n",
              "      <td>PMCID:PMC3940150</td>\n",
              "      <td>efavirenz; non-nucleoside reverse transcriptas...</td>\n",
              "      <td>Allele C is not associated with response to ef...</td>\n",
              "      <td>no</td>\n",
              "      <td>= 0.80</td>\n",
              "      <td>91.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Multiple groups</td>\n",
              "      <td>Efficacy</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>CYP2B6</td>\n",
              "      <td>C</td>\n",
              "      <td>NaN</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5966</th>\n",
              "      <td>1450045235</td>\n",
              "      <td>rs10444413</td>\n",
              "      <td>PMID:29791011</td>\n",
              "      <td>methotrexate</td>\n",
              "      <td>Allele C is not associated with clearance of m...</td>\n",
              "      <td>no</td>\n",
              "      <td>&gt; 0.0001</td>\n",
              "      <td>187.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Unknown</td>\n",
              "      <td>Metabolism/PK</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>SLCO1B1</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2884</th>\n",
              "      <td>982033113</td>\n",
              "      <td>rs10509680</td>\n",
              "      <td>PMID:20833655</td>\n",
              "      <td>warfarin</td>\n",
              "      <td>Allele T is associated with decreased dose of ...</td>\n",
              "      <td>yes</td>\n",
              "      <td>= 3.84E-7</td>\n",
              "      <td>1508.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>East Asian</td>\n",
              "      <td>Dosage</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>CYP2C9</td>\n",
              "      <td>T</td>\n",
              "      <td>G</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2754</th>\n",
              "      <td>981859317</td>\n",
              "      <td>rs10509681</td>\n",
              "      <td>PMCID:PMC3660078</td>\n",
              "      <td>paclitaxel</td>\n",
              "      <td>Allele C is associated with increased risk of ...</td>\n",
              "      <td>yes</td>\n",
              "      <td>= 0.004</td>\n",
              "      <td>411.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Multiple groups, combined European American an...</td>\n",
              "      <td>Toxicity</td>\n",
              "      <td>False</td>\n",
              "      <td>Risk increases per copy of the C allele (indic...</td>\n",
              "      <td>CYP2C8</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2767</th>\n",
              "      <td>1447954627</td>\n",
              "      <td>rs10509681</td>\n",
              "      <td>PMID:24704000</td>\n",
              "      <td>docetaxel; paclitaxel</td>\n",
              "      <td>Allele T is not associated with response to do...</td>\n",
              "      <td>no</td>\n",
              "      <td>= 0.888</td>\n",
              "      <td>22.0</td>\n",
              "      <td>36.0</td>\n",
              "      <td>Central/South Asian</td>\n",
              "      <td>Efficacy</td>\n",
              "      <td>False</td>\n",
              "      <td>Patients with complete and partial pathologica...</td>\n",
              "      <td>CYP2C8</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3387</th>\n",
              "      <td>1184757013</td>\n",
              "      <td>rs9332127</td>\n",
              "      <td>PMCID:PMC4134280</td>\n",
              "      <td>warfarin</td>\n",
              "      <td>Allele G is not associated with dose of warfar...</td>\n",
              "      <td>no</td>\n",
              "      <td>&gt; 0.20</td>\n",
              "      <td>787.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>East Asian</td>\n",
              "      <td>Dosage</td>\n",
              "      <td>False</td>\n",
              "      <td>Samples, genotypes and INRs from a cohort of 5...</td>\n",
              "      <td>CYP2C9</td>\n",
              "      <td>G</td>\n",
              "      <td>C</td>\n",
              "      <td>0.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2939</th>\n",
              "      <td>1447519965</td>\n",
              "      <td>rs9332197</td>\n",
              "      <td>PMCID:PMC3330749</td>\n",
              "      <td>warfarin</td>\n",
              "      <td>Allele C is associated with time to achieve st...</td>\n",
              "      <td>yes</td>\n",
              "      <td>= 3.45e-04</td>\n",
              "      <td>311.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Multiple groups, 98% white</td>\n",
              "      <td>Efficacy</td>\n",
              "      <td>False</td>\n",
              "      <td>NaN</td>\n",
              "      <td>CYP2C9</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3279</th>\n",
              "      <td>1446534354</td>\n",
              "      <td>rs9332238</td>\n",
              "      <td>PMCID:PMC4573240</td>\n",
              "      <td>warfarin</td>\n",
              "      <td>Allele G is associated with increased dose of ...</td>\n",
              "      <td>yes</td>\n",
              "      <td>= 4.4e-13</td>\n",
              "      <td>367.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Multiple groups, Brazilian</td>\n",
              "      <td>Dosage</td>\n",
              "      <td>False</td>\n",
              "      <td>The G allele was strongly associated with high...</td>\n",
              "      <td>CYP2C9</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3905</th>\n",
              "      <td>827788986</td>\n",
              "      <td>rs9332239</td>\n",
              "      <td>PMID:15284535</td>\n",
              "      <td>tolbutamide</td>\n",
              "      <td>Allele T is associated with decreased catalyti...</td>\n",
              "      <td>yes</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Metabolism/PK</td>\n",
              "      <td>False</td>\n",
              "      <td>p value 0.05</td>\n",
              "      <td>CYP2C9</td>\n",
              "      <td>T</td>\n",
              "      <td>NaN</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3897</th>\n",
              "      <td>1451919700</td>\n",
              "      <td>rs9332241</td>\n",
              "      <td>PMCID:PMC9537548</td>\n",
              "      <td>warfarin</td>\n",
              "      <td>Allele T is associated with decreased metaboli...</td>\n",
              "      <td>yes</td>\n",
              "      <td>= 5.38E-10</td>\n",
              "      <td>524.0</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Sub-Saharan African</td>\n",
              "      <td>Metabolism/PK</td>\n",
              "      <td>False</td>\n",
              "      <td>measured as decreased S-warfarin/R-warfarin ra...</td>\n",
              "      <td>CYP2C9</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>1.0</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>278 rows Ã— 17 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-a6d331f4-e225-4c9a-b712-3cbba5babff1')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-a6d331f4-e225-4c9a-b712-3cbba5babff1 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-a6d331f4-e225-4c9a-b712-3cbba5babff1');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 9
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "import requests\n",
        "\n",
        "# Create empty lists to store the genomic information\n",
        "chromosomes = []\n",
        "locations = []\n",
        "reference_bases = []\n",
        "\n",
        "# Base URL for Ensembl REST API\n",
        "base_url = 'https://rest.ensembl.org'\n",
        "\n",
        "# Iterate over rsIDs\n",
        "for rsID in data_unique['Variant']:\n",
        "    # Construct the API endpoint URL\n",
        "    endpoint = f'{base_url}/variation/human/{rsID}?phenotypes=1'\n",
        "\n",
        "    # Send GET request to the API\n",
        "    response = requests.get(endpoint, headers={'Content-Type': 'application/json'})\n",
        "\n",
        "    # Check if the request was successful\n",
        "    if response.status_code == 200:\n",
        "        # Extract relevant information from the response\n",
        "        data = response.json()\n",
        "        chromosome = data['mappings'][0]['seq_region_name']\n",
        "        location = data['mappings'][0]['start']\n",
        "        reference_base = data['mappings'][0]['allele_string'].split('/')[0]\n",
        "\n",
        "        # Append the genomic information to the respective lists\n",
        "        chromosomes.append(chromosome)\n",
        "        locations.append(location)\n",
        "        reference_bases.append(reference_base)\n",
        "    else:\n",
        "        # If there was an error, append None to the lists\n",
        "        chromosomes.append(None)\n",
        "        locations.append(None)\n",
        "        reference_bases.append(None)\n",
        "\n",
        "# Create a new DataFrame with the genomic information\n",
        "search_df = pd.DataFrame({\n",
        "    'Chromosome': chromosomes,\n",
        "    'Location': locations,\n",
        "    'Reference Base': reference_bases\n",
        "})"
      ],
      "metadata": {
        "id": "uRDcicXhaIQz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Combine the search results with orginal dataset\n",
        "search_df.set_index(data_unique.index, inplace=True)\n",
        "combined_df = data_unique.join(search_df)"
      ],
      "metadata": {
        "id": "Mck5jynofKV_"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Filter duplicates\n",
        "\n",
        "combined_df['Combination'] = combined_df[['Alter', 'Ref']].apply(lambda x: ''.join(sorted(x.values.astype(str))), axis=1)\n",
        "# Drop rows where Column1 and Combination values are the same, and Column4 is 'no'\n",
        "filterd_df = combined_df.dropna(subset=['Chromosome','Location','Reference Base'])\n",
        "filterd_df = filterd_df[~((combined_df.duplicated(subset=['Location', 'Combination']) & (combined_df['numeric'] == 0)) |\n",
        "      ((combined_df['numeric'] == 1) & combined_df.duplicated(subset=['Location', 'Combination']) & (combined_df.groupby('Location')['numeric'].transform('max') == 1)))]"
      ],
      "metadata": {
        "id": "p--1R7DmJ-nK",
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "outputId": "5c04e86b-9f05-4f7a-a934-9df6208dd93b"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-12-07e5148f1181>:6: UserWarning: Boolean Series key will be reindexed to match DataFrame index.\n",
            "  filterd_df = filterd_df[~((combined_df.duplicated(subset=['Location', 'Combination']) & (combined_df['numeric'] == 0)) |\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Finalizing\n",
        "#\"Pediatric\" deleted due to irrelavance\n",
        "final1 = filterd_df[['Chromosome','Genes','Location','Reference Base','Alter','Combination','numeric','Literature','Drugs','Association','More Details']]"
      ],
      "metadata": {
        "id": "ZM54ujM4Z-Qz"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Filter again\n",
        "#1)Combination Combination - Reference Base\n",
        "#2)Delete nan Combination\n",
        "#3)Change the rest Combination to \"True Alter\"\n",
        "filtered2 = final1\n",
        "filtered2['Combination'] = final1.apply(lambda row: str(row['Combination']).replace(str(row['Reference Base']), ''), axis=1)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "ISlFuN_RcZsH",
        "outputId": "ac8fa72f-9c79-4db8-fc17-fed2b49c2266"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-14-4d14096c7fa8>:6: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame.\n",
            "Try using .loc[row_indexer,col_indexer] = value instead\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  filtered2['Combination'] = final1.apply(lambda row: str(row['Combination']).replace(str(row['Reference Base']), ''), axis=1)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "data_dropped2 = filtered2[~(filtered2['Combination'] == 'nan')]"
      ],
      "metadata": {
        "id": "t3Ba6V6DJxci"
      },
      "execution_count": null,
      "outputs": []
    },
    {
      "cell_type": "code",
      "source": [
        "# Final 2: Calculate True Alter\n",
        "final2 = data_dropped2\n",
        "final2['Combination'] = data_dropped2.apply(lambda row: row['Alter'] if str(row['Alter']) != str(row['Reference Base']) else str(row['Combination']), axis=1)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "9IrJzXp-M3pD",
        "outputId": "8fd3dda1-6d04-4727-a4cb-9c771b632383"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "stream",
          "name": "stderr",
          "text": [
            "<ipython-input-16-80a9026e924b>:3: SettingWithCopyWarning: \n",
            "A value is trying to be set on a copy of a slice from a DataFrame.\n",
            "Try using .loc[row_indexer,col_indexer] = value instead\n",
            "\n",
            "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
            "  final2['Combination'] = data_dropped2.apply(lambda row: row['Alter'] if str(row['Alter']) != str(row['Reference Base']) else str(row['Combination']), axis=1)\n"
          ]
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "# Final 3: Delete irrelevant column\n",
        "final3 = final2.drop(columns='Alter')\n",
        "final3 = final3.rename(columns={'Combination': 'Altered Base'})\n",
        "final3 = final3[~(final3['Altered Base'] == None)]\n",
        "final3"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 658
        },
        "id": "yCR-uRtTN71V",
        "outputId": "439e529c-0bdd-4816-b78b-5a557b29a042"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "     Chromosome    Genes    Location Reference Base Altered Base  numeric  \\\n",
              "817          19   CYP2B6  41018248.0              T            C      0.0   \n",
              "5966         12  SLCO1B1  21164734.0              T            C      0.0   \n",
              "2884         10   CYP2C9  94974582.0              G            T      1.0   \n",
              "2754         10   CYP2C8  95038992.0              T            C      1.0   \n",
              "3824         10   CYP2C9  94981296.0              A            C      1.0   \n",
              "...         ...      ...         ...            ...          ...      ...   \n",
              "3665         10   CYP2C9  94947714.0              G            C      0.0   \n",
              "2939         10   CYP2C9  94981151.0              T            C      1.0   \n",
              "3279         10   CYP2C9  94988735.0              G            A      1.0   \n",
              "3905         10   CYP2C9  94989020.0              C            T      1.0   \n",
              "3897         10   CYP2C9  94989116.0              C            T      1.0   \n",
              "\n",
              "            Literature                                              Drugs  \\\n",
              "817   PMCID:PMC3940150  efavirenz; non-nucleoside reverse transcriptas...   \n",
              "5966     PMID:29791011                                       methotrexate   \n",
              "2884     PMID:20833655                                           warfarin   \n",
              "2754  PMCID:PMC3660078                                         paclitaxel   \n",
              "3824  PMCID:PMC4134280                                           warfarin   \n",
              "...                ...                                                ...   \n",
              "3665     PMID:22248286                                           warfarin   \n",
              "2939  PMCID:PMC3330749                                           warfarin   \n",
              "3279  PMCID:PMC4573240                                           warfarin   \n",
              "3905     PMID:15284535                                        tolbutamide   \n",
              "3897  PMCID:PMC9537548                                           warfarin   \n",
              "\n",
              "                                            Association  \\\n",
              "817   Allele C is not associated with response to ef...   \n",
              "5966  Allele C is not associated with clearance of m...   \n",
              "2884  Allele T is associated with decreased dose of ...   \n",
              "2754  Allele C is associated with increased risk of ...   \n",
              "3824  Allele A is associated with increased dose of ...   \n",
              "...                                                 ...   \n",
              "3665  Allele C is not associated with increased dose...   \n",
              "2939  Allele C is associated with time to achieve st...   \n",
              "3279  Allele G is associated with increased dose of ...   \n",
              "3905  Allele T is associated with decreased catalyti...   \n",
              "3897  Allele T is associated with decreased metaboli...   \n",
              "\n",
              "                                           More Details  \n",
              "817                                                 NaN  \n",
              "5966                                                NaN  \n",
              "2884                                                NaN  \n",
              "2754  Risk increases per copy of the C allele (indic...  \n",
              "3824  Samples, genotypes and INRs from a cohort of 5...  \n",
              "...                                                 ...  \n",
              "3665  No association was found between this variant ...  \n",
              "2939                                                NaN  \n",
              "3279  The G allele was strongly associated with high...  \n",
              "3905                                       p value 0.05  \n",
              "3897  measured as decreased S-warfarin/R-warfarin ra...  \n",
              "\n",
              "[240 rows x 10 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-c4f3c995-8527-4f62-8ed0-8116e8a6f882\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Chromosome</th>\n",
              "      <th>Genes</th>\n",
              "      <th>Location</th>\n",
              "      <th>Reference Base</th>\n",
              "      <th>Altered Base</th>\n",
              "      <th>numeric</th>\n",
              "      <th>Literature</th>\n",
              "      <th>Drugs</th>\n",
              "      <th>Association</th>\n",
              "      <th>More Details</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>817</th>\n",
              "      <td>19</td>\n",
              "      <td>CYP2B6</td>\n",
              "      <td>41018248.0</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMCID:PMC3940150</td>\n",
              "      <td>efavirenz; non-nucleoside reverse transcriptas...</td>\n",
              "      <td>Allele C is not associated with response to ef...</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5966</th>\n",
              "      <td>12</td>\n",
              "      <td>SLCO1B1</td>\n",
              "      <td>21164734.0</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMID:29791011</td>\n",
              "      <td>methotrexate</td>\n",
              "      <td>Allele C is not associated with clearance of m...</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2884</th>\n",
              "      <td>10</td>\n",
              "      <td>CYP2C9</td>\n",
              "      <td>94974582.0</td>\n",
              "      <td>G</td>\n",
              "      <td>T</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMID:20833655</td>\n",
              "      <td>warfarin</td>\n",
              "      <td>Allele T is associated with decreased dose of ...</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2754</th>\n",
              "      <td>10</td>\n",
              "      <td>CYP2C8</td>\n",
              "      <td>95038992.0</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMCID:PMC3660078</td>\n",
              "      <td>paclitaxel</td>\n",
              "      <td>Allele C is associated with increased risk of ...</td>\n",
              "      <td>Risk increases per copy of the C allele (indic...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3824</th>\n",
              "      <td>10</td>\n",
              "      <td>CYP2C9</td>\n",
              "      <td>94981296.0</td>\n",
              "      <td>A</td>\n",
              "      <td>C</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMCID:PMC4134280</td>\n",
              "      <td>warfarin</td>\n",
              "      <td>Allele A is associated with increased dose of ...</td>\n",
              "      <td>Samples, genotypes and INRs from a cohort of 5...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3665</th>\n",
              "      <td>10</td>\n",
              "      <td>CYP2C9</td>\n",
              "      <td>94947714.0</td>\n",
              "      <td>G</td>\n",
              "      <td>C</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMID:22248286</td>\n",
              "      <td>warfarin</td>\n",
              "      <td>Allele C is not associated with increased dose...</td>\n",
              "      <td>No association was found between this variant ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>2939</th>\n",
              "      <td>10</td>\n",
              "      <td>CYP2C9</td>\n",
              "      <td>94981151.0</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMCID:PMC3330749</td>\n",
              "      <td>warfarin</td>\n",
              "      <td>Allele C is associated with time to achieve st...</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3279</th>\n",
              "      <td>10</td>\n",
              "      <td>CYP2C9</td>\n",
              "      <td>94988735.0</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMCID:PMC4573240</td>\n",
              "      <td>warfarin</td>\n",
              "      <td>Allele G is associated with increased dose of ...</td>\n",
              "      <td>The G allele was strongly associated with high...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3905</th>\n",
              "      <td>10</td>\n",
              "      <td>CYP2C9</td>\n",
              "      <td>94989020.0</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMID:15284535</td>\n",
              "      <td>tolbutamide</td>\n",
              "      <td>Allele T is associated with decreased catalyti...</td>\n",
              "      <td>p value 0.05</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>3897</th>\n",
              "      <td>10</td>\n",
              "      <td>CYP2C9</td>\n",
              "      <td>94989116.0</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMCID:PMC9537548</td>\n",
              "      <td>warfarin</td>\n",
              "      <td>Allele T is associated with decreased metaboli...</td>\n",
              "      <td>measured as decreased S-warfarin/R-warfarin ra...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>240 rows Ã— 10 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-c4f3c995-8527-4f62-8ed0-8116e8a6f882')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-c4f3c995-8527-4f62-8ed0-8116e8a6f882 button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-c4f3c995-8527-4f62-8ed0-8116e8a6f882');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 17
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "np.shape(final3)"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/"
        },
        "id": "FdQxQ11hUqEm",
        "outputId": "34542972-dbb1-4934-fd73-2638186147fe"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "(240, 10)"
            ]
          },
          "metadata": {},
          "execution_count": 18
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "final4 = final3.sort_values(['Chromosome', 'Genes'])\n",
        "final4"
      ],
      "metadata": {
        "colab": {
          "base_uri": "https://localhost:8080/",
          "height": 658
        },
        "id": "FMbPyF1CjpKN",
        "outputId": "12390cfd-d348-4a00-8d8a-726ae8b9f21a"
      },
      "execution_count": null,
      "outputs": [
        {
          "output_type": "execute_result",
          "data": {
            "text/plain": [
              "     Chromosome          Genes    Location Reference Base Altered Base  \\\n",
              "206          10          ABCC2  99861014.0              C            G   \n",
              "135          10          ABCC2  99852563.0              C            T   \n",
              "34           10          ABCC2  99843765.0              T            C   \n",
              "46           10          ABCC2  99804028.0              C            T   \n",
              "37           10          ABCC2  99804266.0              C            T   \n",
              "...         ...            ...         ...            ...          ...   \n",
              "5594          7         CYP3A4  99760901.0              A            G   \n",
              "5706          7         CYP3A4  99758184.0          TTTTT            T   \n",
              "5737          7         CYP3A4  99761528.0              C            T   \n",
              "5645          7  CYP3A; CYP3A4  99762047.0              G            A   \n",
              "5742          7  CYP3A; CYP3A4  99760836.0              G            A   \n",
              "\n",
              "      numeric        Literature  \\\n",
              "206       1.0     PMID:18294295   \n",
              "135       1.0     PMID:27180762   \n",
              "34        1.0     PMID:30100615   \n",
              "46        1.0     PMID:30100615   \n",
              "37        0.0     PMID:21449672   \n",
              "...       ...               ...   \n",
              "5594      1.0     PMID:12966368   \n",
              "5706      0.0  PMCID:PMC7235792   \n",
              "5737      1.0  PMCID:PMC3237759   \n",
              "5645      1.0     PMID:11470997   \n",
              "5742      0.0     PMID:15634941   \n",
              "\n",
              "                                                  Drugs  \\\n",
              "206                                           docetaxel   \n",
              "135                                         vincristine   \n",
              "34                                         rosuvastatin   \n",
              "46                                         rosuvastatin   \n",
              "37                                       antiepileptics   \n",
              "...                                                 ...   \n",
              "5594                                         tacrolimus   \n",
              "5706  ambrisentan; aripiprazole; atorvastatin; donep...   \n",
              "5737                                         tacrolimus   \n",
              "5645                                                NaN   \n",
              "5742                                         nifedipine   \n",
              "\n",
              "                                            Association  \\\n",
              "206   Allele G is associated with increased risk of ...   \n",
              "135   Allele T is associated with decreased likeliho...   \n",
              "34    Allele C is associated with decreased exposure...   \n",
              "46    Allele T is associated with decreased exposure...   \n",
              "37    Allele C is not associated with resistance to ...   \n",
              "...                                                 ...   \n",
              "5594  Allele G is associated with decreased dose of ...   \n",
              "5706  Allele T is not associated with metabolism of ...   \n",
              "5737  Allele C is associated with tacrolimus concent...   \n",
              "5645  Allele A is associated with decreased steady-s...   \n",
              "5742  Allele A is not associated with decreased meta...   \n",
              "\n",
              "                                           More Details  \n",
              "206   Study genotyped 84 patients from BioBank Japan...  \n",
              "135   when comparing zero neurotoxicity vs grades 1-...  \n",
              "34                                                  NaN  \n",
              "46                                                  NaN  \n",
              "37    Epilepsy patients on antiepileptic drug (AED) ...  \n",
              "...                                                 ...  \n",
              "5594  as calculated by increased tacrolimus dose-adj...  \n",
              "5706  No significant changes in PK parameters in the...  \n",
              "5737  This variant is in linkage disequilibrium with...  \n",
              "5645  In a bacterial heterologous expression system,...  \n",
              "5742  In vitro.  Metabolism was not significantly di...  \n",
              "\n",
              "[240 rows x 10 columns]"
            ],
            "text/html": [
              "\n",
              "  <div id=\"df-3dabfab4-af5c-4310-a23d-7985e6db069a\">\n",
              "    <div class=\"colab-df-container\">\n",
              "      <div>\n",
              "<style scoped>\n",
              "    .dataframe tbody tr th:only-of-type {\n",
              "        vertical-align: middle;\n",
              "    }\n",
              "\n",
              "    .dataframe tbody tr th {\n",
              "        vertical-align: top;\n",
              "    }\n",
              "\n",
              "    .dataframe thead th {\n",
              "        text-align: right;\n",
              "    }\n",
              "</style>\n",
              "<table border=\"1\" class=\"dataframe\">\n",
              "  <thead>\n",
              "    <tr style=\"text-align: right;\">\n",
              "      <th></th>\n",
              "      <th>Chromosome</th>\n",
              "      <th>Genes</th>\n",
              "      <th>Location</th>\n",
              "      <th>Reference Base</th>\n",
              "      <th>Altered Base</th>\n",
              "      <th>numeric</th>\n",
              "      <th>Literature</th>\n",
              "      <th>Drugs</th>\n",
              "      <th>Association</th>\n",
              "      <th>More Details</th>\n",
              "    </tr>\n",
              "  </thead>\n",
              "  <tbody>\n",
              "    <tr>\n",
              "      <th>206</th>\n",
              "      <td>10</td>\n",
              "      <td>ABCC2</td>\n",
              "      <td>99861014.0</td>\n",
              "      <td>C</td>\n",
              "      <td>G</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMID:18294295</td>\n",
              "      <td>docetaxel</td>\n",
              "      <td>Allele G is associated with increased risk of ...</td>\n",
              "      <td>Study genotyped 84 patients from BioBank Japan...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>135</th>\n",
              "      <td>10</td>\n",
              "      <td>ABCC2</td>\n",
              "      <td>99852563.0</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMID:27180762</td>\n",
              "      <td>vincristine</td>\n",
              "      <td>Allele T is associated with decreased likeliho...</td>\n",
              "      <td>when comparing zero neurotoxicity vs grades 1-...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>34</th>\n",
              "      <td>10</td>\n",
              "      <td>ABCC2</td>\n",
              "      <td>99843765.0</td>\n",
              "      <td>T</td>\n",
              "      <td>C</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMID:30100615</td>\n",
              "      <td>rosuvastatin</td>\n",
              "      <td>Allele C is associated with decreased exposure...</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>46</th>\n",
              "      <td>10</td>\n",
              "      <td>ABCC2</td>\n",
              "      <td>99804028.0</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMID:30100615</td>\n",
              "      <td>rosuvastatin</td>\n",
              "      <td>Allele T is associated with decreased exposure...</td>\n",
              "      <td>NaN</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>37</th>\n",
              "      <td>10</td>\n",
              "      <td>ABCC2</td>\n",
              "      <td>99804266.0</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMID:21449672</td>\n",
              "      <td>antiepileptics</td>\n",
              "      <td>Allele C is not associated with resistance to ...</td>\n",
              "      <td>Epilepsy patients on antiepileptic drug (AED) ...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>...</th>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "      <td>...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5594</th>\n",
              "      <td>7</td>\n",
              "      <td>CYP3A4</td>\n",
              "      <td>99760901.0</td>\n",
              "      <td>A</td>\n",
              "      <td>G</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMID:12966368</td>\n",
              "      <td>tacrolimus</td>\n",
              "      <td>Allele G is associated with decreased dose of ...</td>\n",
              "      <td>as calculated by increased tacrolimus dose-adj...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5706</th>\n",
              "      <td>7</td>\n",
              "      <td>CYP3A4</td>\n",
              "      <td>99758184.0</td>\n",
              "      <td>TTTTT</td>\n",
              "      <td>T</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMCID:PMC7235792</td>\n",
              "      <td>ambrisentan; aripiprazole; atorvastatin; donep...</td>\n",
              "      <td>Allele T is not associated with metabolism of ...</td>\n",
              "      <td>No significant changes in PK parameters in the...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5737</th>\n",
              "      <td>7</td>\n",
              "      <td>CYP3A4</td>\n",
              "      <td>99761528.0</td>\n",
              "      <td>C</td>\n",
              "      <td>T</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMCID:PMC3237759</td>\n",
              "      <td>tacrolimus</td>\n",
              "      <td>Allele C is associated with tacrolimus concent...</td>\n",
              "      <td>This variant is in linkage disequilibrium with...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5645</th>\n",
              "      <td>7</td>\n",
              "      <td>CYP3A; CYP3A4</td>\n",
              "      <td>99762047.0</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>1.0</td>\n",
              "      <td>PMID:11470997</td>\n",
              "      <td>NaN</td>\n",
              "      <td>Allele A is associated with decreased steady-s...</td>\n",
              "      <td>In a bacterial heterologous expression system,...</td>\n",
              "    </tr>\n",
              "    <tr>\n",
              "      <th>5742</th>\n",
              "      <td>7</td>\n",
              "      <td>CYP3A; CYP3A4</td>\n",
              "      <td>99760836.0</td>\n",
              "      <td>G</td>\n",
              "      <td>A</td>\n",
              "      <td>0.0</td>\n",
              "      <td>PMID:15634941</td>\n",
              "      <td>nifedipine</td>\n",
              "      <td>Allele A is not associated with decreased meta...</td>\n",
              "      <td>In vitro.  Metabolism was not significantly di...</td>\n",
              "    </tr>\n",
              "  </tbody>\n",
              "</table>\n",
              "<p>240 rows Ã— 10 columns</p>\n",
              "</div>\n",
              "      <button class=\"colab-df-convert\" onclick=\"convertToInteractive('df-3dabfab4-af5c-4310-a23d-7985e6db069a')\"\n",
              "              title=\"Convert this dataframe to an interactive table.\"\n",
              "              style=\"display:none;\">\n",
              "        \n",
              "  <svg xmlns=\"http://www.w3.org/2000/svg\" height=\"24px\"viewBox=\"0 0 24 24\"\n",
              "       width=\"24px\">\n",
              "    <path d=\"M0 0h24v24H0V0z\" fill=\"none\"/>\n",
              "    <path d=\"M18.56 5.44l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94zm-11 1L8.5 8.5l.94-2.06 2.06-.94-2.06-.94L8.5 2.5l-.94 2.06-2.06.94zm10 10l.94 2.06.94-2.06 2.06-.94-2.06-.94-.94-2.06-.94 2.06-2.06.94z\"/><path d=\"M17.41 7.96l-1.37-1.37c-.4-.4-.92-.59-1.43-.59-.52 0-1.04.2-1.43.59L10.3 9.45l-7.72 7.72c-.78.78-.78 2.05 0 2.83L4 21.41c.39.39.9.59 1.41.59.51 0 1.02-.2 1.41-.59l7.78-7.78 2.81-2.81c.8-.78.8-2.07 0-2.86zM5.41 20L4 18.59l7.72-7.72 1.47 1.35L5.41 20z\"/>\n",
              "  </svg>\n",
              "      </button>\n",
              "      \n",
              "  <style>\n",
              "    .colab-df-container {\n",
              "      display:flex;\n",
              "      flex-wrap:wrap;\n",
              "      gap: 12px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert {\n",
              "      background-color: #E8F0FE;\n",
              "      border: none;\n",
              "      border-radius: 50%;\n",
              "      cursor: pointer;\n",
              "      display: none;\n",
              "      fill: #1967D2;\n",
              "      height: 32px;\n",
              "      padding: 0 0 0 0;\n",
              "      width: 32px;\n",
              "    }\n",
              "\n",
              "    .colab-df-convert:hover {\n",
              "      background-color: #E2EBFA;\n",
              "      box-shadow: 0px 1px 2px rgba(60, 64, 67, 0.3), 0px 1px 3px 1px rgba(60, 64, 67, 0.15);\n",
              "      fill: #174EA6;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert {\n",
              "      background-color: #3B4455;\n",
              "      fill: #D2E3FC;\n",
              "    }\n",
              "\n",
              "    [theme=dark] .colab-df-convert:hover {\n",
              "      background-color: #434B5C;\n",
              "      box-shadow: 0px 1px 3px 1px rgba(0, 0, 0, 0.15);\n",
              "      filter: drop-shadow(0px 1px 2px rgba(0, 0, 0, 0.3));\n",
              "      fill: #FFFFFF;\n",
              "    }\n",
              "  </style>\n",
              "\n",
              "      <script>\n",
              "        const buttonEl =\n",
              "          document.querySelector('#df-3dabfab4-af5c-4310-a23d-7985e6db069a button.colab-df-convert');\n",
              "        buttonEl.style.display =\n",
              "          google.colab.kernel.accessAllowed ? 'block' : 'none';\n",
              "\n",
              "        async function convertToInteractive(key) {\n",
              "          const element = document.querySelector('#df-3dabfab4-af5c-4310-a23d-7985e6db069a');\n",
              "          const dataTable =\n",
              "            await google.colab.kernel.invokeFunction('convertToInteractive',\n",
              "                                                     [key], {});\n",
              "          if (!dataTable) return;\n",
              "\n",
              "          const docLinkHtml = 'Like what you see? Visit the ' +\n",
              "            '<a target=\"_blank\" href=https://colab.research.google.com/notebooks/data_table.ipynb>data table notebook</a>'\n",
              "            + ' to learn more about interactive tables.';\n",
              "          element.innerHTML = '';\n",
              "          dataTable['output_type'] = 'display_data';\n",
              "          await google.colab.output.renderOutput(dataTable, element);\n",
              "          const docLink = document.createElement('div');\n",
              "          docLink.innerHTML = docLinkHtml;\n",
              "          element.appendChild(docLink);\n",
              "        }\n",
              "      </script>\n",
              "    </div>\n",
              "  </div>\n",
              "  "
            ]
          },
          "metadata": {},
          "execution_count": 19
        }
      ]
    },
    {
      "cell_type": "code",
      "source": [
        "final4.to_csv('/content/gdrive/My Drive/Colab Notebooks/ADME Project/ListedCore_auto.csv')"
      ],
      "metadata": {
        "id": "Q8SY8WmRivqO"
      },
      "execution_count": null,
      "outputs": []
    }
  ]
}